### PR TITLE
Change agent code to force usage of SSL from the beginning

### PR DESCRIPTION
**Pull request by ketan**
Most of this code is around maintaining backward compatibility,
upgrading of commons httpclient from 3.0.1 to httpcomponents 4.5.2

### CODE REVIEW

File: installers/agent/win/config/wrapper-agent.conf [hunk 1]
```diff
@@ -62,8 +62,8 @@ set.AGENT_STARTUP_ARGS=-Xms128m -Xmx256m %JVM_DEBUG% %GC_LOG% "-DJAVA_SYS_MON_TE
 
 # Application parameters.  Add parameters as needed starting from 1
 wrapper.app.parameter.1=com.simontuffs.onejar.Boot
-wrapper.app.parameter.2=%GO_SERVER%
-wrapper.app.parameter.3=8153
+wrapper.app.parameter.2=-serverUrl
```
Comment (jyotisingh) [inline installers/agent/win/config/wrapper-agent.conf]: If users have either of these two properties overridden in their `wrapper-properties.conf`, the *new* values from wrapper-agent.conf will be ignored. Meaning, there could be users running Go server on port 8000, so they would have an entry `wrapper.app.parameter.3=8000` in their wrapper-properties.conf. Upon upgrade, bootstrapper expects `GO_SERVER_URL` to be passed along but now we would get a port number instead. Either we should handle it (meaning log proper error if this happens), or have some sort of warning somewhere (in docs/release notes?) stating the same.

File: installers/shared/shared_post.sh [hunk 2]
```diff
@@ -143,8 +143,10 @@ function create_necessary_agent_directories {
 }
 
 function print_agent_configuration_suggestions {
-    echo "Now please edit $GO_AGENT_DEFAULTS and set GO_SERVER to the IP address of your Go Server."
-    echo "Once that is done start the Go Agent with '/etc/init.d/go-agent start'"
+    echo "Now please edit $GO_AGENT_DEFAULTS and set GO_SERVER_URL to the url address of your GoCD Server (https://example.com:8154/go)."
+    echo "To ensure full end-to-end transport security between the agent and server"
+    echo "see https://docs.go.cd/current/installation/ssl_tls_config.html#end-to-end-transport-security."
```
Comment (jyotisingh) [inline installers/shared/shared_post.sh]: wrong docs url? that url points to a chapter heading page

File: agent-bootstrapper/src/com/thoughtworks/go/agent/bootstrapper/AgentBootstrapper.java [hunk 3]
```diff
@@ -46,24 +48,19 @@ public AgentBootstrapper(AgentLauncherCreator launcherCreator) {
         this.launcherCreator = launcherCreator;
     }
 
-    public static void main(String[] args) {
-        BootstrapperLoggingHelper.initLog4j();
```
Comment (jyotisingh) [inline agent-bootstrapper/src/com/thoughtworks/go/agent/bootstrapper/AgentBootstrapper.java]: This used to setup logging for agent-bootstrapper. Is getting rid of this intentional?

File: agent-bootstrapper/src/com/thoughtworks/go/agent/bootstrapper/DefaultAgentLaunchDescriptorImpl.java [hunk 4]
```diff
@@ -1,47 +1,43 @@
...
-    public DefaultAgentLaunchDescriptorImpl(String hostname, int port, AgentBootstrapper self) {
-        this.bootstrapper = self;
-        buildContext(hostname, port);
+    public DefaultAgentLaunchDescriptorImpl(AgentBootstrapperArgs bootstrapperArgs, AgentBootstrapper agentBootstrapper) {
+        this.bootstrapper = agentBootstrapper;
+        buildContext(bootstrapperArgs);
     }
 
-    private void buildContext(String hostname, int port) {
-        context.putAll(System.getenv());
```
Comment (jyotisingh) [inline agent-bootstrapper/src/com/thoughtworks/go/agent/bootstrapper/DefaultAgentLaunchDescriptorImpl.java]: `AgentBootstrapperArgs.toProperties()` does not include `System.getenv()` and `System.getProperties()`. What are the repercussions of removing these? Can we do it as a part of a different PR instead and possibly add a toggle around it too?

File: installers/agent/release/agent.sh [hunk 5]
```diff
@@ -63,14 +63,36 @@ location of your Java installation."
...
+declare -a _stringToArgs
+function stringToArgsArray() {
+  _stringToArgs=("$@")
+}
+
+function autoDetectGoServerUrl() {
+  local url
+
+  if [[ ! -z "${GO_SERVER}" || ! -z "${GO_SERVER_PORT}" ]]; then
+    yell "The environment variable GO_SERVER and GO_SERVER_PORT has been deprecated in favor of GO_SERVER_URL. Please set GO_SERVER_URL instead to a https url (https://example.com:8154/go)"
```
Comment (jyotisingh) [inline installers/agent/release/agent.sh]: Could you mark it as a `WARN` since the agent continues to run even after this message. Also, the message is echoed to console which could be missed if there is no user interaction while starting the service. Could we echo this to the stdout file too.

**Comment by jyotisingh** [comment]
I was also trying out a few scenarios myself, found an issue with one of them. 
If I have a fresh 16.4 server with 16.4 remote agent, agent has been approved etc. Now I delete the server keystore(along with agentkeystore), then restart the server, the agent starts getting the following error:
```
2016-06-01 13:41:01,525 [loopThread] ERROR thoughtworks.go.agent.AgentController:203 - There has be
en a problem with one of Go's SSL certificates. This can be caused by a man-in-the-middle attack, o
r by pointing the agent to a new server, or by deleting and re-installing Go Server. Go will ask fo
r a new certificate. If this fails to solve the problem, try deleting config/trust.jks in Go Agent'
s home directory.
javax.net.ssl.SSLHandshakeException: sun.security.validator.ValidatorException: PKIX path validatio
n failed: java.security.cert.CertPathValidatorException: signature check failed
        at sun.security.ssl.Alerts.getSSLException(Alerts.java:192)
        at sun.security.ssl.SSLSocketImpl.fatal(SSLSocketImpl.java:1916)
        at sun.security.ssl.Handshaker.fatalSE(Handshaker.java:279)
        at sun.security.ssl.Handshaker.fatalSE(Handshaker.java:273)
        at sun.security.ssl.ClientHandshaker.serverCertificate(ClientHandshaker.java:1472)
        at sun.security.ssl.ClientHandshaker.processMessage(ClientHandshaker.java:213)
        at sun.security.ssl.Handshaker.processLoop(Handshaker.java:913)
        at sun.security.ssl.Handshaker.process_record(Handshaker.java:849)
        at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:1035)
        at sun.security.ssl.SSLSocketImpl.performInitialHandshake(SSLSocketImpl.java:1344)
        at sun.security.ssl.SSLSocketImpl.writeRecord(SSLSocketImpl.java:721)
        at sun.security.ssl.AppOutputStream.write(AppOutputStream.java:122)
        at java.io.BufferedOutputStream.flushBuffer(BufferedOutputStream.java:82)
        at java.io.BufferedOutputStream.flush(BufferedOutputStream.java:140)
        at org.apache.commons.httpclient.HttpConnection.flushRequestOutputStream(HttpConnection.jav
a:827)
        at org.apache.commons.httpclient.MultiThreadedHttpConnectionManager$HttpConnectionAdapter.f
lushRequestOutputStream(MultiThreadedHttpConnectionManager.java:1525)
        at org.apache.commons.httpclient.HttpMethodBase.writeRequest(HttpMethodBase.java:1975)
        at org.apache.commons.httpclient.HttpMethodBase.execute(HttpMethodBase.java:993)
        at org.apache.commons.httpclient.HttpMethodDirector.executeWithRetry(HttpMethodDirector.jav
a:397)
        at org.apache.commons.httpclient.HttpMethodDirector.executeMethod(HttpMethodDirector.java:1
70)
        at org.apache.commons.httpclient.HttpClient.executeMethod(HttpClient.java:396)
        at org.apache.commons.httpclient.HttpClient.executeMethod(HttpClient.java:324)
        at com.thoughtworks.go.agent.service.AgentUpgradeService.checkForUpgrade(AgentUpgradeService.java:54)
        at com.thoughtworks.go.agent.service.AgentUpgradeService.checkForUpgrade(AgentUpgradeService.java:47)
        at com.thoughtworks.go.agent.AgentController.rpcLoop(AgentController.java:172)
        at com.thoughtworks.go.agent.AgentController.loop(AgentController.java:165)
```
but it eventually recovers from it, and the agent continues to be registered with the server:
```
2016-06-01 13:41:11,633 [loopThread] INFO  go.agent.service.SslInfrastructureService:93 - [Agent Registration] Starting to register agent
2016-06-01 13:41:17,037 [loopThread] INFO  go.agent.service.SslInfrastructureService:124 - [Agent Registration] Retrieved registration from Go server.
2016-06-01 13:41:17,038 [loopThread] INFO  go.agent.service.SslInfrastructureService:131 - [Agent Registration] Stored registration for cert with hash code: fCLtAetQa+1fB4BnjOIKuA== not valid before: Thu Jan 01 05:30:00 IST 1970
2016-06-01 13:41:17,039 [loopThread] INFO  go.agent.service.SslInfrastructureService:97 - [Agent Registration] Successfully registered agent
```

When, I try the same with fresh 16.6 server with a fresh 16.6 agent (both built from this PR) with verification-mode set to `NONE` and try the same scenario, the agent gets stuck at the following error and does not recover:
```
2016-06-01 13:55:33,317 [pingThread] ERROR thoughtworks.go.agent.AgentController:155 - Error occurred when agent tried to ping server:
org.springframework.remoting.RemoteAccessException: Could not access HTTP invoker remote service at [https://10.4.3.209:8154/go/remoting/remoteBuildRepository]; nested exception is javax.net.ssl.SSLHandshakeException: Remote host closed connection during handshake
	at org.springframework.remoting.httpinvoker.HttpInvokerClientInterceptor.convertHttpInvokerAccessException(HttpInvokerClientInterceptor.java:212)
	at org.springframework.remoting.httpinvoker.HttpInvokerClientInterceptor.invoke(HttpInvokerClientInterceptor.java:145)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:172)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:202)
	at com.sun.proxy.$Proxy6.ping(Unknown Source)
	at com.thoughtworks.go.agent.AgentController.ping(AgentController.java:151)
	at sun.reflect.GeneratedMethodAccessor5.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:606)
	at org.springframework.util.MethodInvoker.invoke(MethodInvoker.java:273)
	at org.springframework.scheduling.support.MethodInvokingRunnable.run(MethodInvokingRunnable.java:65)
	at org.springframework.scheduling.timer.DelegatingTimerTask.run(DelegatingTimerTask.java:70)
	at java.util.TimerThread.mainLoop(Timer.java:555)
	at java.util.TimerThread.run(Timer.java:505)
Caused by: javax.net.ssl.SSLHandshakeException: Remote host closed connection during handshake
	at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:965)
	at sun.security.ssl.SSLSocketImpl.performInitialHandshake(SSLSocketImpl.java:1344)
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1371)
	at sun.security.ssl.SSLSocketImpl.startHandshake(SSLSocketImpl.java:1355)
	at org.apache.http.conn.ssl.SSLConnectionSocketFactory.createLayeredSocket(SSLConnectionSocketFactory.java:394)
	at org.apache.http.conn.ssl.SSLConnectionSocketFactory.connectSocket(SSLConnectionSocketFactory.java:353)
	at org.apache.http.impl.conn.DefaultHttpClientConnectionOperator.connect(DefaultHttpClientConnectionOperator.java:141)
	at org.apache.http.impl.conn.PoolingHttpClientConnectionManager.connect(PoolingHttpClientConnectionManager.java:353)
	at org.apache.http.impl.execchain.MainClientExec.establishRoute(MainClientExec.java:380)
	at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:236)
	at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:184)
	at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:88)
	at org.apache.http.impl.execchain.RedirectExec.execute(RedirectExec.java:110)
	at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:184)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:82)
	at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:107)
	at com.thoughtworks.go.agent.common.ssl.GoAgentServerHttpClient.execute(GoAgentServerHttpClient.java:47)
	at com.thoughtworks.go.agent.GoHttpClientHttpInvokerRequestExecutor.doExecuteRequest(GoHttpClientHttpInvokerRequestExecutor.java:51)
	at org.springframework.remoting.httpinvoker.AbstractHttpInvokerRequestExecutor.executeRequest(AbstractHttpInvokerRequestExecutor.java:136)
	at org.springframework.remoting.httpinvoker.HttpInvokerClientInterceptor.executeRequest(HttpInvokerClientInterceptor.java:192)
	at org.springframework.remoting.httpinvoker.HttpInvokerClientInterceptor.executeRequest(HttpInvokerClientInterceptor.java:174)
	at org.springframework.remoting.httpinvoker.HttpInvokerClientInterceptor.invoke(HttpInvokerClientInterceptor.java:142)
	... 12 more
Caused by: java.io.EOFException: SSL peer shut down incorrectly
	at sun.security.ssl.InputRecord.read(InputRecord.java:482)
	at sun.security.ssl.SSLSocketImpl.readRecord(SSLSocketImpl.java:946)
	... 33 more
```

File: installers/agent/win/config/wrapper-agent.conf (hunk 1 above)
Comment (ketan) [inline installers/agent/win/config/wrapper-agent.conf]: > Either we should handle it (meaning log proper error if this happens), or have some sort of warning somewhere (in docs/release notes?) stating the same

Not sure how we can "handle it", in such a scenario, the bootstrapper will refuse to startup and print the usage onto the console.

Will make sure to submit a PR for release notes to document this behaviour.

File: agent-bootstrapper/src/com/thoughtworks/go/agent/bootstrapper/DefaultAgentLaunchDescriptorImpl.java (hunk 4 above)
Comment (ketan) [inline agent-bootstrapper/src/com/thoughtworks/go/agent/bootstrapper/DefaultAgentLaunchDescriptorImpl.java]: I checked thoroughly for usages of `AgentLaunchDescriptor#context` on master, and found that those keys are just not required or used. The only keys that ware used were `hostname` and `port`.

File: installers/agent/release/agent.sh (hunk 5 above)
Comment (ketan) [inline installers/agent/release/agent.sh]: Changing `yell` to print `WARN: message` and `die` to print `FATAL: message`

File: installers/shared/shared_post.sh (hunk 2 above)
Comment (ketan) [inline installers/shared/shared_post.sh]: Updated url (to a non-existant page, pending a PR on the docs repo)

**Comment by ketan** [comment]
Patch to allow usage of http proxy system variables —

```patch
diff --git a/base/src/com/thoughtworks/go/agent/common/ssl/GoAgentServerHttpClientBuilder.java b/base/src/com/thoughtworks/go/agent/common/ssl/GoAgentServerHttpClientBuilder.java
index 6464737..be9ca59 100644
--- a/base/src/com/thoughtworks/go/agent/common/ssl/GoAgentServerHttpClientBuilder.java
+++ b/base/src/com/thoughtworks/go/agent/common/ssl/GoAgentServerHttpClientBuilder.java
@@ -25,10 +25,13 @@ import org.apache.http.impl.client.CloseableHttpClient;
 import org.apache.http.impl.client.DefaultConnectionKeepAliveStrategy;
 import org.apache.http.impl.client.HttpClientBuilder;
 import org.apache.http.impl.client.HttpClients;
+import org.apache.http.impl.conn.DefaultSchemePortResolver;
+import org.apache.http.impl.conn.SystemDefaultRoutePlanner;
 import org.apache.http.ssl.SSLContextBuilder;
 
 import javax.net.ssl.HostnameVerifier;
 import java.io.*;
+import java.net.ProxySelector;
 import java.security.*;
 import java.security.cert.CertificateException;
 import java.security.cert.X509Certificate;
@@ -91,6 +94,9 @@ public class GoAgentServerHttpClientBuilder {
 
         SSLConnectionSocketFactory sslsf = new SSLConnectionSocketFactory(sslContextBuilder.build(), hostnameVerifier);
         builder.setSSLSocketFactory(sslsf);
+
+        builder.setRoutePlanner(new SystemDefaultRoutePlanner(DefaultSchemePortResolver.INSTANCE, ProxySelector.getDefault()));
+
         return builder;
     }
 
```
